### ISSUE TITLE
URIs not Equivalent with Differently Ordered Query Strings

### ISSUE THREAD

**Issue by oalders** [libwww-perl/URI#35]
See https://rt.cpan.org/Ticket/Display.html?id=81491

I'm wondering if rather than

```perl
sub _obj_eq {
    return overload::StrVal($_[0]) eq overload::StrVal($_[1]);
}
```

we should do something like

```perl
sub _obj_eq {
    return $_[0]->canonical->as_string eq $_[1]->canonical->as_string;
}
```

I haven't tested this, but wanted to start a discussion here.

Ping @eserte and @theory who are commenters on the original RT.

**Comment by theory** [comment]
@srezic had a valid point in the original RT that order might be important in determining equivalence. He said:

> I think order matters here. Somebody could use a query string like
> 
>     foo.bar?ordered_array=1;ordered_array=2;ordered_array=3;ordered_array=4
> 
> and expect it to be ordered.

So I can imagine sometimes wanting it to be considered equivalent and sometimes not, depending on the use-case and context. Maybe it'd make sense to add a method that compares without regard to query string ordering?

**Comment by oalders** [comment]
Is there a use case where the order of parameters actually matters?  (I haven't read the RFC).

**Comment by genio** [comment]
Yes, order matters. Mojo::Parameters's example should likely be followed.

```
11:19 <@sri> if you change order the query strings become semantically different according to the specs
11:19 <@sri> which may or may not be important, depending on the apps involved
```

**Comment by oalders** [comment]
> Note: this reply was drafted with the help of an LLM and reviewed before posting.

Coming back to this after a long while. To ground the discussion in the spec: RFC 3986 treats the query as opaque, "non-hierarchical data" (§3.4) — it notes that `key=value` pairs are *commonly* used but deliberately doesn't define that structure as part of the generic syntax. And none of its defined normalizations (§6.2.2: case, percent-encoding, path-segment) reorder the query. So under the generic spec, two URIs differing only in query-parameter order are **not** equivalent; any order-insensitive notion of equality is explicitly left to application- or scheme-specific semantics (§6.2.3).

That matches the current behavior: `eq` compares `canonical->as_string`, and `canonical` doesn't reorder query parameters, so `?a=1&b=2` and `?b=2&a=1` correctly compare as **not equal**. (The originally-proposed switch of `_obj_eq` to `canonical` was never adopted — and wouldn't achieve order-insensitivity anyway, since it's deliberately an object-identity check behind `==`, and `canonical` doesn't sort params.)

So the default is working as intended, and I'm closing this. The one idea still worth having is the earlier suggestion in this thread of an **opt-in, order-insensitive comparison** helper — that remains welcome as a PR for anyone who wants it.
